File: filters/suggestions_to_forum.py
import discord
from discord.ext import commands

SUGGESTION_CHANNEL_ID = 1441660601460850708
PRIORITIES_FORUM_ID   = 1441702171136626768
MESSAGE_FETCH_LIMIT   = 50
MIN_REACTIONS         = 1
WHITE_CHECK = "✅"   
PROCESSED_MARK = "📌"  


# -----------------------------
# CHANNEL FETCHING & VALIDATION
# -----------------------------
def get_suggestion_channels(bot: commands.Bot):
    suggestion = bot.get_channel(SUGGESTION_CHANNEL_ID)
    priorities = bot.get_channel(PRIORITIES_FORUM_ID)

    errors = []
    if not isinstance(suggestion, discord.TextChannel):
        errors.append(f"Suggestion channel (ID: {SUGGESTION_CHANNEL_ID}) missing or not a TextChannel")
    if not isinstance(priorities, discord.ForumChannel):
        errors.append(f"Priorities forum (ID: {PRIORITIES_FORUM_ID}) missing or not a ForumChannel")

    return suggestion, priorities, errors


# -----------------------------
# DELETE ALL THREADS IN FORUM
# -----------------------------
async def clear_forum_threads(forum_channel: discord.ForumChannel):
    print("[INFO] Clearing priorities forum...")

    active_threads = forum_channel.threads
    archived_threads = [t async for t in forum_channel.archived_threads(limit=None)]
    all_threads = active_threads + archived_threads

    deleted = 0

    for thread in all_threads:
        try:
            await thread.delete()
            deleted += 1
        except Exception as e:
            print(f"[ERROR] Failed to delete thread {thread.id}: {e}")

    print(f"[INFO] Deleted {deleted} threads.")


# -----------------------------
# GET CHECKMARK COUNT
# -----------------------------
def get_checkmark_count(msg: discord.Message) -> int:
    for reaction in msg.reactions:
        if str(reaction.emoji) == WHITE_CHECK:
            return reaction.count
    return 0


# -----------------------------
# FETCH, FILTER & SORT MESSAGES
# -----------------------------
async def fetch_filtered_suggestions(channel: discord.TextChannel, limit: int):
    messages = [msg async for msg in channel.history(limit=limit)]

    # Filter for MIN_REACTIONS or above
    filtered = [
        msg for msg in messages
        if get_checkmark_count(msg) >= MIN_REACTIONS
    ]

    # Sort by reaction count (highest → lowest)
    filtered.sort(key=get_checkmark_count, reverse=True)

    return filtered

# -----------------------------
# PARSE MESSAGE CONTENT
# -----------------------------
def extract_suggestion_fields(msg: discord.Message):
    lines = msg.content.split("\n")

    title = ""
    content_lines = []
    found_title = False

    for line in lines:
        stripped = line.strip()

        # Look for "Title : something"
        if stripped.lower().startswith("title"):
            # Split only once on the first colon
            parts = stripped.split(":", 1)
            if len(parts) > 1:
                title = parts[1].strip()
                found_title = True
            continue

        # Everything after title is part of content
        if found_title:
            content_lines.append(line)

    content = "\n".join(content_lines).strip()

    return title, content


# -----------------------------
# ATTACHMENT HANDLING
# -----------------------------
async def get_attachments(msg: discord.Message):
    files = []
    for attachment in msg.attachments:
        if attachment.content_type and attachment.content_type.startswith("image"):
            files.append(await attachment.to_file())
    return files


# -----------------------------
# POST THREAD
# -----------------------------
async def post_priority_thread(forum_channel, title, content, files):
    try:
        await forum_channel.create_thread(
            name=title,
            content=content,
            files=files
        )
        return True
    except Exception as e:
        print(f"[ERROR] Failed to create thread '{title}': {e}")
        return False


# -----------------------------
# MAIN FUNCTION
# -----------------------------
async def post_suggestions_to_priorities(bot: commands.Bot, limit: int = MESSAGE_FETCH_LIMIT):

    suggestion_channel, priorities_forum, errors = get_suggestion_channels(bot)

    if errors:
        for err in errors:
            print(f"[ERROR] {err}")
        print("[INFO] Stopping because channels are missing.")
        return

    # Clear forum before adding new posts
    await clear_forum_threads(priorities_forum)

    # Fetch valid suggestions
    suggestions = await fetch_filtered_suggestions(suggestion_channel, limit)
    posted = 0

    for msg in suggestions:
        title, description = extract_suggestion_fields(msg)
        if not title:
            continue  # Must have a title

        content = f"{description}\n\n{msg.jump_url}"
        files = await get_attachments(msg)

        if await post_priority_thread(priorities_forum, title, content, files):
            posted += 1
            try:
                await msg.add_reaction("📌")
            except Exception as e:
                print(f"[ERROR] Failed to add processed mark to message {msg.id}: {e}")

            # PIN the message
            try:
                if not msg.pinned:  # Only pin if not already pinned
                    await msg.pin(reason="Synced by bot")
            except Exception as e:
                print(f"[ERROR] Failed to pin message {msg.id}: {e}")



    print(f"[INFO] Posted {posted} suggestions to priorities.")

File: filters/suggestions_to_forum.js
const {
    ForumChannel,
    TextChannel,
    AttachmentBuilder,
    ChannelType
} = require("discord.js");

const { getServerConfig } = require("../databases/servers");

const MESSAGE_FETCH_LIMIT = 50;
const WHITE_CHECK = "✅";
const MINIMUM_REACTIONS = 1;

/* ---------------------------------------------------------
   SAFE CHANNEL FETCHER
--------------------------------------------------------- */

/**
 * Safely fetch and validate a channel.
 * Handles: missing channel, missing access, wrong type.
 * @param {Client} client
 * @param {string} channelId
 * @param {ChannelType} expectedType
 * @returns {Promise<{channel: any, error: string|null}>}
 */
async function safeGetChannel(client, channelId, expectedType) {
    try {
        const channel = await client.channels.fetch(channelId).catch(() => null);

        if (!channel) {
            return {
                channel: null,
                error: `[Missing] Channel ${channelId} does not exist or bot cannot access it`
            };
        }

        if (channel.type !== expectedType) {
            return {
                channel: null,
                error: `[Invalid] Channel ${channelId} is not type ${expectedType}`
            };
        }

        return { channel, error: null };
    } catch (err) {
        return {
            channel: null,
            error: `[Error] Cannot fetch channel ${channelId}: ${err.message}`
        };
    }
}

/* ---------------------------------------------------------
   CHANNEL RETRIEVAL
--------------------------------------------------------- */

async function getSuggestionChannels(client, guildId) {
    const config = await getServerConfig(guildId);

    if (!config) {
        return {
            suggestion: null,
            priorities: null,
            errors: [`Server ${guildId} is not configured in database`]
        };
    }

    const errors = [];

    const { channel: suggestion, error: e1 } =
        await safeGetChannel(client, config.SUGGESTION_CHANNEL_ID, ChannelType.GuildText);

    const { channel: priorities, error: e2 } =
        await safeGetChannel(client, config.PRIORITIES_SUGGESTION_FORUM_ID, ChannelType.GuildForum);

    if (e1) errors.push(e1);
    if (e2) errors.push(e2);

    return { suggestion, priorities, errors };
}

/* ---------------------------------------------------------
   CLEAR FORUM THREADS (SAFE)
--------------------------------------------------------- */

async function clearForumThreads(forum) {
    if (!forum) {
        console.log("[ERROR] Cannot clear forum: Forum is null");
        return;
    }

    console.log("[INFO] Clearing priorities forum...");

    try {
        const active = await forum.threads.fetchActive().catch(() => ({ threads: new Map() }));
        const archived = await forum.threads.fetchArchived().catch(() => ({ threads: new Map() }));

        const allThreads = [
            ...active.threads.values(),
            ...archived.threads.values()
        ];

        let deletedCount = 0;

        for (const thread of allThreads) {
            try {
                await thread.delete();
                deletedCount++;
            } catch (err) {
                console.log("[ERROR] Failed deleting thread:", err.message);
            }
        }

        console.log(`[INFO] Deleted ${deletedCount} threads.`);
    } catch (err) {
        console.log("[ERROR] Unable to clear forum threads:", err.message);
    }
}

/* ---------------------------------------------------------
   SUGGESTION UTILITIES
--------------------------------------------------------- */

function getCheckmarkCount(msg) {
    const r = msg.reactions.cache.find(r => r.emoji.name === WHITE_CHECK);
    return r ? r.count : 0;
}

async function fetchSuggestions(channel, limit) {
    try {
        const messages = await channel.messages.fetch({ limit });
        const arr = [...messages.values()];

        arr.sort((a, b) => {
            const ra = getCheckmarkCount(a);
            const rb = getCheckmarkCount(b);
            return ra !== rb ? rb - ra : a.createdTimestamp - b.createdTimestamp;
        });

        return arr;
    } catch (err) {
        console.log("[ERROR] Cannot fetch messages:", err.message);
        return [];
    }
}

function extractSuggestionFields(msg) {
    const lines = msg.content.split("\n");

    for (let i = 0; i < Math.min(4, lines.length); i++) {
        const line = lines[i].trim();
        const match = line.match(/^(?:[^a-zA-Z\u0600-\u06FF]*)(title|titre|العنوان)\s*:\s*(.+)$/i);

        if (match) {
            const title = match[2].trim();
            const content = lines.slice(i + 1).join("\n").trim();
            return { title, content };
        }
    }
    return null;
}

async function getAttachments(msg) {
    const attachments = [];

    for (const a of msg.attachments.values()) {
        if (a.contentType?.startsWith("image")) {
            try {
                const response = await fetch(a.url);
                const buffer = Buffer.from(await response.arrayBuffer());
                attachments.push(new AttachmentBuilder(buffer, { name: a.name }));
            } catch (err) {
                console.log("[ERROR] Failed downloading attachment:", err.message);
            }
        }
    }

    return attachments;
}

/* ---------------------------------------------------------
   CREATE THREAD SAFELY
--------------------------------------------------------- */

async function postPriorityThread(forum, title, content, files) {
    try {
        await forum.threads.create({
            name: title,
            message: { content, files }
        });
        return true;
    } catch (err) {
        console.log("[ERROR] Failed creating thread:", err.message);
        return false;
    }
}

/* ---------------------------------------------------------
   MAIN FUNCTION
--------------------------------------------------------- */

async function postSuggestionsToPriorities(client, guildId, limit = MESSAGE_FETCH_LIMIT) {

    const { suggestion, priorities, errors } = await getSuggestionChannels(client, guildId);

    if (errors.length > 0) {
        console.log("[ERROR] Channel configuration issues:");
        errors.forEach(e => console.log("  -", e));
        return;
    }

    // Clear old threads
    await clearForumThreads(priorities);

    // Fetch suggestions
    const suggestions = await fetchSuggestions(suggestion, limit);

    let postedCount = 0;

    for (const msg of suggestions.reverse()) {
        if (getCheckmarkCount(msg) < MINIMUM_REACTIONS) continue;

        const fields = extractSuggestionFields(msg);
        if (!fields) continue;

        const finalText = `${fields.content}\n\n${msg.url}`;
        const files = await getAttachments(msg);

        const success = await postPriorityThread(priorities, fields.title, finalText, files);
        if (success) postedCount++;
    }

    console.log(`[INFO] Posted ${postedCount} suggestions for server ${guildId}.`);
}

module.exports = { postSuggestionsToPriorities };
